### ISSUE TITLE
Lua engine issues with sol3

### ISSUE THREAD

**Issue by cuavas** [mamedev/mame#7485]
We need to sort out any remaining Lua issues before we can release MAME 0.227.

~~Uncommenting the `device` property for the `image` user type causes Sol to attempt to instantiate `device_t` which obviously won’t work.  Are we supposed to use `new_usertype<device_t &>` (i.e. with a reference type) to indicate it’s a reference rather than value type?~~

The [usertype documentation](https://sol2.readthedocs.io/en/latest/api/usertype.html) suggests we should perhaps use `("device", sol::no_constructor)` (i.e. without the `"new"` argument) to create a type without a constructor.

~~There seem to be some issues with value types.  Perhaps we’re not supposed to be using the `sol::user<T>` wrapper when returning things or passing them to Sol functions?~~

Need to check that the custom `sol_lua_get`/`sol_lua_check`/`sol_lua_push` are being found by ADL properly.  Also, does the `buffer` class need to be in the `sol` namespace now?

**Comment by cuavas** [comment]
An example of something going wrong with a value type:
```
[MAME]> ipt = manager:machine():input()
[MAME]> p = manager:machine():ioport().ports
[MAME]> seq = p[":P1_P2"].fields["P1 Up"]:default_input_seq("standard")
[MAME]> n = ipt:seq_name(seq)
```
The final line crashes with an access violation.  The function is defined like this:
```
input_type.set("seq_name", [](input_manager &input, const input_seq &seq) { return input.seq_name(seq); });
```
But it’s possible something is going wrong earlier, leaving `seq` with a dangling reference.

The crash dump looks like this:
```
#0  0x00000000088ca702 in input_manager::seq_clean(input_seq const&) const ()
#1  0x00000000088cab59 in input_manager::seq_name[abi:cxx11](input_seq const&) const ()
#2  0x00000000050c5ee9 in lua_engine::initialize()::$_149::__invoke[abi:cxx11](input_manager&, input_seq const&) ()
#3  0x0000000005231dc9 in int sol::stack::call_into_lua<false, true, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, , input_manager&, input_seq const&, sol::wrapper<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > (*)(input_manager&, input_seq const&), void>::caller, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > (*&)(input_manager&, input_seq const&)>(sol::types<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >>, sol::types<input_manager&, input_seq const&>, lua_State*, int, sol::wrapper<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > (*)(input_manager&, input_seq const&), void>::caller&&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > (*&)(input_manager&, input_seq const&))
    ()
#4  0x00000000050c5f30 in int sol::u_detail::binding<char [9], lua_engine::initialize()::$_149(input_manager)>::call_<true, false>(lua_State*) ()
#5  0x0000000005111dae in sol::detail::lua_cfunction_trampoline(lua_State*, int (*)(lua_State*)) ()
#6  0x000000000926e2bb in luaD_precall ()
#7  0x0000000009283e98 in luaV_execute ()
#8  0x000000000926e4d6 in luaD_callnoyield ()
#9  0x000000000926d6ce in luaD_rawrunprotected ()
#10 0x000000000926eacb in luaD_pcall ()
#11 0x00000000092694a0 in lua_pcallk ()
#12 0x0000000009287500 in luaB_pcall ()
#13 0x000000000926e2bb in luaD_precall ()
#14 0x0000000009283e98 in luaV_execute ()
#15 0x000000000926e4d6 in luaD_callnoyield ()
#16 0x000000000926d6ce in luaD_rawrunprotected ()
#17 0x000000000926eacb in luaD_pcall ()
#18 0x00000000092694a0 in lua_pcallk ()
#19 0x0000000005110456 in sol::protected_function_result sol::basic_protected_function<sol::basic_reference<false>, false, sol::basic_reference<false> >::invoke<false>(sol::types<>, std::integer_sequence<unsigned long>, long, sol::detail::protected_handler<false, sol::basic_reference<false> >&) const ()
#20 0x0000000005110bf2 in decltype(auto) sol::basic_protected_function<sol::basic_reference<false>, false, sol::basic_reference<false> >::call<>() const ()
#21 0x00000000050645a7 in std::_Function_handler<bool (sol::basic_protected_function<sol::basic_reference<false>, false, sol::basic_reference<false> > const&), lua_engine::execute_function(char const*)::$_0>::_M_invoke(std::_Any_data const&, sol::basic_protected_function<sol::basic_reference<false>, false, sol::basic_reference<false> > const&) ()
#22 0x0000000004fefaea in lua_engine::enumerate_functions(char const*, std::function<bool (sol::basic_protected_function<sol::basic_reference<false>, false, sol::basic_reference<false> > const&)>&&) ()
#23 0x0000000004ff06a8 in lua_engine::on_periodic() ()
#24 0x00000000089aacbb in video_manager::frame_update(bool) ()
#25 0x00000000088fbad0 in running_machine::run(bool) ()
#26 0x0000000004f46159 in mame_machine_manager::execute() ()
#27 0x0000000004fcb7e2 in cli_frontend::start_execution(mame_machine_manager*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&) ()
#28 0x0000000004fcd0fb in cli_frontend::execute(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&) ()
#29 0x0000000004f4706c in emulator_info::start_frontend(emu_options&, osd_interface&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&) ()
#30 0x0000000004e7f980 in main ()
```

**Comment by cuavas** [comment]
It’s crashing pretty close to the start of the function:
```
#0  0x00000000088ca702 in input_manager::seq_clean(input_seq const&) const ()
#1  0x00000000088cab59 in input_manager::seq_name[abi:cxx11](input_seq const&) const ()
#2  0x00000000050c5ee9 in lua_engine::initialize()::$_149::__invoke[abi:cxx11](input_manager&, input_seq const&) ()
#3  0x0000000005231dc9 in int sol::stack::call_into_lua<false, true, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, , input_manager&, input_seq const&, sol::wrapper<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > (*)(input_manager&, input_seq const&), void>::caller, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > (*&)(input_manager&, input_seq const&)>(sol::types<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >>, sol::types<input_manager&, input_seq const&>, lua_State*, int, sol::wrapper<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > (*)(input_manager&, input_seq const&), void>::caller&&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > (*&)(input_manager&, input_seq const&))
    ()
#4  0x00000000050c5f30 in int sol::u_detail::binding<char [9], lua_engine::initialize()::$_149(input_manager)>::call_<true, false>(lua_State*) ()
#5  0x0000000005111dae in sol::detail::lua_cfunction_trampoline(lua_State*, int (*)(lua_State*)) ()
#6  0x000000000926e2bb in luaD_precall ()
#7  0x0000000009283e98 in luaV_execute ()
#8  0x000000000926e4d6 in luaD_callnoyield ()
#9  0x000000000926d6ce in luaD_rawrunprotected ()
#10 0x000000000926eacb in luaD_pcall ()
#11 0x00000000092694a0 in lua_pcallk ()
#12 0x0000000009287500 in luaB_pcall ()
#13 0x000000000926e2bb in luaD_precall ()
#14 0x0000000009283e98 in luaV_execute ()
#15 0x000000000926e4d6 in luaD_callnoyield ()
#16 0x000000000926d6ce in luaD_rawrunprotected ()
#17 0x000000000926eacb in luaD_pcall ()
#18 0x00000000092694a0 in lua_pcallk ()
#19 0x0000000005110456 in sol::protected_function_result sol::basic_protected_function<sol::basic_reference<false>, false, sol::basic_reference<false> >::invoke<false>(sol::types<>, std::integer_sequence<unsigned long>, long, sol::detail::protected_handler<false, sol::basic_reference<false> >&) const ()
#20 0x0000000005110bf2 in decltype(auto) sol::basic_protected_function<sol::basic_reference<false>, false, sol::basic_reference<false> >::call<>() const ()
#21 0x00000000050645a7 in std::_Function_handler<bool (sol::basic_protected_function<sol::basic_reference<false>, false, sol::basic_reference<false> > const&), lua_engine::execute_function(char const*)::$_0>::_M_invoke(std::_Any_data const&, sol::basic_protected_function<sol::basic_reference<false>, false, sol::basic_reference<false> > const&) ()
#22 0x0000000004fefaea in lua_engine::enumerate_functions(char const*, std::function<bool (sol::basic_protected_function<sol::basic_reference<false>, false, sol::basic_reference<false> > const&)>&&) ()
#23 0x0000000004ff06a8 in lua_engine::on_periodic() ()
#24 0x00000000089aacbb in video_manager::frame_update(bool) ()
#25 0x00000000088fbad0 in running_machine::run(bool) ()
#26 0x0000000004f46159 in mame_machine_manager::execute() ()
#27 0x0000000004fcb7e2 in cli_frontend::start_execution(mame_machine_manager*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&) ()
#28 0x0000000004fcd0fb in cli_frontend::execute(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&) ()
#29 0x0000000004f4706c in emulator_info::start_frontend(emu_options&, osd_interface&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&) ()
#30 0x0000000004e7f980 in main ()
```

R14 was loaded from RDX, which is the third argument including the hidden return value location and `this` pointer.  It has value `0x500000ea10010052` in this case, so it isn’t a null pointer, but it’s not valid.

**Comment by cuavas** [comment]
It works without `sol::make_user` everywhere.  As of 3cb183a6b283df67b050bb5cf108a8d5e752c09c I can use the autofire plugin.  I don’t know if there are any downsides to this.

**Comment by cuavas** [comment]
The device thing is fixed in 7e7b5f488eac3592c982e9d7d1266157e98a0519 – `const` is no longer cast away.

**Comment by cuavas** [comment]
Now might also be a good time to think about whether we can split up luaengine.cpp into smaller pieces.  Is it really necessary to have everything in one file?  What if we moved e.g. the render stuff and input stuff to different files to hopefully cut down on memory required to build it?  We’d need a .ipp file for the customised get/push/check, but apart from that, should it work in theory?

**Comment by cuavas** [comment]
I’ve split it into a few pieces now.  Should be a bit more manageable.

**Comment by cuavas** [comment]
The `sol::buffer` has gone away and all the ADL tricks seem to be working.  Closing as done.
